ols/src/prompts/prompt_generator.py

```diff
@@ -1,78 +1,126 @@
-    rag_context: str,
-) -> tuple[PromptTemplate, dict[str, Any]]:
-    """Dynamically creates prompt template and input values specification for LLM.
-
-    Prompt template is created by combining these values:
-    - provider
-    - model
-    - model-specific configuration
-    - RAG (if enabled and exists)
-    - history
-    - system prompt
-    - user query
-    """
-    # Add system instruction to the prompt first.
-    selector = PromptConfiguration(provider, model)
-    prompt = prompt_configurations.get(selector, QUERY_SYSTEM_INSTRUCTION)
-    llm_input_values = {"query": query}
-
-    # Add optional instruction to the prompt & create prompt inputs.
-    if len(rag_context) > 0:
-        prompt += USE_CONTEXT_INSTRUCTION
-        prompt += CONTEXT_PLACEHOLDER
-
-        llm_input_values["context"] = rag_context
-
-    if len(history) > 0:
-        prompt += USE_HISTORY_INSTRUCTION
-        prompt += HISTORY_PLACEHOLDER
-        llm_input_values["chat_history"] = "\n".join(history)
-
-    prompt += QUERY_PLACEHOLDER
-
-    return PromptTemplate.from_template(prompt), llm_input_values
+
+def restructure_rag_context_pre(text: str, model: str) -> str:
+    """Restructure rag text - pre truncation."""
+    if ModelFamily.GRANITE in model:
+        return "\n[End]\n[Document]\n" + text
+    else:
+        return "\n\nDocument:\n" + text
+
+
+def restructure_rag_context_post(text: str, model: str) -> str:
+    """Restructure rag text - post truncation."""
+    if ModelFamily.GRANITE in model:
+        return text.removeprefix("\n[End]") + "\n[End]"
+    else:
+        return "\n" + text.lstrip("\n") + "\n"
+
+
+def restructure_history(text: str, model: str) -> str:
+    """Restructure history."""
+    if text.startswith("human: "):
+        if ModelFamily.GRANITE in model:
+            return "\n<|user|>\n" + text.removeprefix("human: ")
+    else:
+        if ModelFamily.GRANITE in model:
+            return "\n<|assistant|>\n" + text.removeprefix("ai: ")
+
+    # No processing required here for gpt.
+    return text
+
+
+class GeneratePrompt:
+    """Generate prompt dynamically."""
+
+    def __init__(
+        self,
+        query: str,
+        rag_context: list[str] = [],
+        history: list[str] = [],
+        system_instruction: str = QUERY_SYSTEM_INSTRUCTION,
+    ):
+        """Initialize prompt generator."""
+        self._query = query
+        self._rag_context = rag_context
+        self._history = history
+        self._sys_instruction = system_instruction
+
+    def _generate_prompt_gpt(self) -> tuple[ChatPromptTemplate, dict]:
+        """Generate prompt for GPT."""
+        prompt_message = []
+        sys_intruction = self._sys_instruction.strip() + "\n"
+        llm_input_values: dict = {"query": self._query}
+
+        if len(self._rag_context) > 0:
+            llm_input_values["context"] = "".join(self._rag_context)
+            sys_intruction = sys_intruction + "\n" + USE_CONTEXT_INSTRUCTION.strip()
```

Comment:
@asamal4 Do we need this \n here ? we already have in the line 69 - may be needed , trying to understand .